---
title: "01_Process snATAC-seq with ArchR"
output: html_notebook
---

```{r message=F}
library(ArchR);library(cowplot);library(Seurat)
set.seed(123)
addArchRThreads(threads = 20)
```

# Read fragments files and generate arrow files
```{r}
addArchRGenome("mm10")
inputFiles <- c("../JL20001/E12_fragments.tsv.gz","../JL20002/E13_fragments.tsv.gz","../JL20003/E14_fragments.tsv.gz")
ArrowFiles <- createArrowFiles(
  inputFiles = inputFiles,
  sampleNames = c("E12", "E13", "E14"),
  filterTSS = 5, 
  filterFrags = 1250,
  maxFrags = 1e+07,
  addTileMat = TRUE,
  addGeneScoreMat = TRUE,
  force = FALSE
)
```

# Infer doublets
For this dataset, we eventually decided not to remove doublets.
```{r}
doubScores <- addDoubletScores(
  input = ArrowFiles,
  k = 10, 
  knnMethod = "UMAP", #Refers to the embedding to use for nearest neighbor search.
  LSIMethod = 1
)
```

# Initialize an ArchRProject
```{r}
proj <- ArchRProject(
  ArrowFiles = ArrowFiles, 
  outputDirectory = "E12_14",
  copyArrows = TRUE 
)

length(proj$cellNames)

# # Filtering Doublets from an ArchRProject
# proj <- filterDoublets(ArchRProj = proj)
# length(proj$cellNames)
```

# Perform dimensionality reduction and clustering
```{r}
proj <- addIterativeLSI(
  ArchRProj = proj,
  useMatrix = "TileMatrix", 
  name = "IterativeLSI", 
  iterations = 2, 
  clusterParams = list( #See Seurat::FindClusters
    resolution = c(0.2), 
    sampleCells = 10000, 
    n.start = 10
  ), 
  varFeatures = 25000, 
  dimsToUse = 1:35,
  force = TRUE
)  

proj <- addClusters(
  input = proj,
  reducedDims = "IterativeLSI",
  method = "Seurat",
  name = "Clusters",
  resolution = 0.8,
  force = TRUE
)

# Visualizing in a 2D UMAP Embedding
proj <- addUMAP(
  ArchRProj = proj, 
  reducedDims = "IterativeLSI", 
  name = "UMAP", 
  nNeighbors = 30, 
  minDist = 0.3, 
  metric = "cosine",
  force = TRUE
)

p1 <- plotEmbedding(ArchRProj = proj, colorBy = "cellColData", name = "Clusters", embedding = "UMAP")
p2 <- plotEmbedding(ArchRProj = proj, colorBy = "cellColData", name = "Sample", embedding = "UMAP")

# plotPDF(p1,p2, name = "Plot-UMAP-Clusters_new.pdf", ArchRProj = proj, addDOC = FALSE, width = 10, height = 5)
```

# Find feature genes using gene score matrix
```{r}
markersGS <- getMarkerFeatures(
  ArchRProj = proj, 
  useMatrix = "GeneScoreMatrix", 
  groupBy = "Clusters",
  bias = c("TSSEnrichment", "log10(nFrags)"),
  testMethod = "wilcoxon"
)

markerList <- getMarkers(markersGS, cutOff = "FDR <= 0.01 & Log2FC >= 1.25")

sig_genes <- data.frame(unlist(markerList))
sig_genes$cluster <- rownames(sig_genes)
# openxlsx::write.xlsx(sig_genes, "E12_14/results/marker.xlsx")
```

# Annotate the cell clusters in excel and add the cell type information to the ArchR object
```{r}
sig_genes <- openxlsx::read.xlsx("E12_14/results/marker_annotated.xlsx", sheet = 2)
tab <- sig_genes[!duplicated(sig_genes$cluster),]
clusters <- tab$cluster
cellType <- tab$cellType
proj$cellType1 <- proj$Clusters
proj$cellType1 <- mapLabels(proj$Clusters, newLabels = cellType, oldLabels = clusters)

p1 <- plotEmbedding(proj, name = "Clusters", embedding = "UMAP", size = 0.1, labelAsFactors=F, labelMeans=T)+NoLegend()+NoAxes()
p2 <- plotEmbedding(proj, name = "cellType1", embedding = "UMAP", size = 0.1, labelAsFactors=F, labelMeans=T)+NoLegend()+NoAxes()

plot_grid(p1,p2)
```

# Make pseudo-bulk replicates and call peaks with MACS2
```{r}
proj <- addGroupCoverages(ArchRProj = proj, groupBy = "cellType1", force = FALSE)

pathToMacs2 <- "/home/CAM/jali/anaconda3/envs/python27/bin/macs2"
proj <- addReproduciblePeakSet(
  ArchRProj = proj,
  groupBy = "cellType1",
  pathToMacs2 = pathToMacs2,
  force = FALSE
)
```

# Add peak matrix and save object
```{r}
proj <- addPeakMatrix(proj)
getAvailableMatrices(proj)

# Save object
saveArchRProject(
  ArchRProj = proj,
  outputDirectory = "E12_14",
  overwrite = TRUE,
  load = FALSE
)
```

